New linux sockets

ranechita opened the pull request:
Add support for linux sockets. iio-demo can now be run and tested without a physical board.

mchindri commented on iio/iio_app/iio_app.c:
```diff
@@ -180,10 +193,19 @@ int32_t iio_app_run(struct iio_app_device *devices, int32_t len)
 	iio_init_param.phy_type = USE_UART;
 	iio_init_param.uart_init_param = &uart_init_par;
-#endif //USE_TCP_SOCKET
+#endif
```
It is useful to have a comment at the end of the endif to know which ended

mchindri commented on network/linux_s/linux_s.c:
```diff
@@ -0,0 +1,300 @@
+/***************************************************************************//**
+ * @file linux_s.c
```
yes, I think linux_socket will be a better naming

mchindri commented on network/linux_s/linux_s.c:
```diff
@@ -0,0 +1,300 @@
+ * @return
+ *  - \ref SUCCESS : On success
+ */
+int32_t linux_get_network_interface(struct linux_desc *desc,
```
You can just use extern for linux_net  in linux_s.h like this iio example.
https://github.com/analogdevicesinc/no-OS/tree/master/iio/iio_ad9083

I was using wifi_get_network_interface because I needed to intialize the .net value.

mchindri commented on network/linux_s/linux_s.c:
```diff
@@ -0,0 +1,300 @@
+
+
+
+struct network_interface linux_net = {
```
If you move the structure initialization at the end of the file you won't need to declare all functions.

mchindri commented on network/linux_s/linux_s.c:
```diff
@@ -0,0 +1,300 @@
+				 enum socket_protocol prot, uint32_t buff_size)
+{
+	*sock_id = socket(AF_INET,SOCK_STREAM,IPPROTO_TCP);
+	int flags = fcntl(*sock_id, F_GETFL);
```
A good practice is to have the variables declarations at the beginning of the function.

mchindri replied:
Same for below functions :)

mchindri commented on network/linux_s/linux_s.c:
```diff
@@ -0,0 +1,300 @@
+	int flags = fcntl(*sock_id, F_GETFL);
+	fcntl(*sock_id, F_SETFL, flags | O_NONBLOCK);
+
+	if(*sock_id < 0)
```
Is better to check this before calling fcntl.

mchindri commented on network/linux_s/linux_s.c:
```diff
@@ -0,0 +1,300 @@
+				    struct socket_address *addr)
+{
+	int32_t ret;
+	struct sockaddr_in saddr;
```
struct sockaddr_in saddr = { 0 };
will not need memset to be called

mchindri commented on network/linux_s/linux_s.c:
```diff
@@ -0,0 +1,300 @@
+
+	if(*sock_id < 0)
+		return *sock_id;
+	return SUCCESS;
```
coding style:  Add new line before return.

mchindri commented on network/linux_s/linux_s.c:
```diff
@@ -0,0 +1,300 @@
+	int32_t ret;
+
+	int flags = fcntl(sock_id, F_GETFL);
+	fcntl(sock_id, F_SETFL, flags | O_NONBLOCK);
```
Do a call to fcntl is needs to be done in each send and recv ? 
Isn't enough in open and accept? Because is the same sock_id.

mchindri commented on network/linux_s/linux_s.c:
```diff
@@ -0,0 +1,300 @@
+				   const void *data, uint32_t size,
+				   const struct socket_address* to)
+{
+	int32_t ret;
```
Did you tested udp socket implementations?

mchindri commented on network/linux_s/linux_s.h:
```diff
@@ -0,0 +1,101 @@
+
+int32_t linux_get_network_interface(struct linux_desc *desc,
+				    struct network_interface** net);
+
```
if you remove linux_get_network_interface, in this file you will just need:
extern struct network_interface linux_net;

mchindri commented on projects/iio_demo/src/app/main.c:
```diff
@@ -223,7 +223,7 @@ int main(void)
 	adc_init_par = (struct adc_demo_init_param) {
 		.loopback_buffers = NULL,
-		.channel_no = 2,
+		.channel_no = 3,
```
Changes in main.c are not related to this commit

mchindri commented:
When doing a force push add a description with the changes you have made:
like: https://github.com/analogdevicesinc/no-OS/pull/958#issuecomment-793849101

Also you need to rebase onto master an resolve the new conflict.

ranechita commented:
v2

- removed additional spaces
- removed get_linux_network interface, replaced it with `extern struct network_interface linux_net`
- removed comment in error.h
- fixed some styling issues

ranechita commented:
v3

- modified Makefile.linux and added $(NO-OS)/tools/scripts/iio_srcs.mk
-  removed comments from src.mk and unnecessary files

ranechita commented on util/xml.c:
```diff
@@ -43,6 +43,7 @@
 
 #include <string.h>
 #include <stdlib.h>
+#include <stdint.h>
```
it was included in one of the files and it wouldn't let me build without it

ranechita commented:
v4

- rename linux_s to linux_socket
- update src.mk
- add iio-app in separate commit

mchindri commented on include/error.h:
```diff
@@ -57,8 +57,10 @@
  * user errors
  */
+#define __ELASTERROR 2000
 #define EOVERRUN	(__ELASTERROR + 1) /* Circular buffer overrun */
```
I would still go with my suggestion. 
Or use:
#ifndef __ELASTERROR 
#define __ELASTERROR 2000
#endif

Otherwise in some platforms a warning with __ELASTERROR already defined will show

mchindri commented on util/xml.c:
```diff
@@ -43,6 +43,7 @@
 
 #include <string.h>
 #include <stdlib.h>
+#include <stdint.h>
```
In commit titles:
utils:xml.c Fix compile issue.

mchindri commented on projects/iio_demo/Makefile.linux:
```diff
@@ -0,0 +1,39 @@
+
+EXEC = iio_demo_linux
```
Fix commit message

mchindri commented on network/linux_socket/linux_socket.h:
```diff
@@ -0,0 +1,88 @@
+
+#define SOCKET_NO 4000
+#define SOCKETS (SOCKET_NO + 1)
+
+/******************************************************************************/
+/*************************** Types Declarations *******************************/
+/******************************************************************************/
+
+struct linux_desc {
+	/*Id of the open socket*/
+	uint32_t id;
+
+	/*Sockets*/
+	int32_t sockets[SOCKETS];
+
+	/*Server descriptor*/
+	int32_t server;
+
+	/*Reference to the network*/
+	struct network_interface net;
+};
+
+struct linux_init_param {
+	/*	Maximum size for incoming data
+		uint32_t max_buff_size;*/
+
+	/* Reference to the network*/
+	struct network_interface* net;
+};
```
This is not needed. init function was removed

mchindri commented on network/linux_socket/linux_socket.h:
```diff
@@ -0,0 +1,88 @@
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *******************************************************************************/
+
+#ifndef NOOS_NETWORK_LINUX_S_LINUX_S_H_
```
#ifndef LINUX_SOCKET_H

mchindri commented on network/linux_socket/linux_socket.h:
```diff
@@ -0,0 +1,88 @@
+/***************************************************************************//**
+ * @file linux_s.h
```
Fix file name here too

mchindri reviewed (commented):
Please check previous comment I didn't mark as resolved

ranechita commented:
v5

- fix commit names
- added ifdef for _ELASTERROR in error.h
- deleted linux_init_param frm linux_socket.h

ranechita commented:
v6

- removed the code for setting non-blocking flags from send and recv and kept it only in accept and open

adrimbarean commented on projects/iio_demo/Makefile.linux:
```diff
@@ -0,0 +1,39 @@
+include ./src.mk
+include  $(NO-OS)/tools/scripts/iio_srcs.mk
+
+all: copy $(EXEC)
+
+copy:
+	mkdir -p $(BUILD_DIR)
+	cp -r $(SRCS) $(BUILD_DIR)
+	cp -r $(INCS) $(BUILD_DIR)
+	cp -r $(SRC_DIRS) $(BUILD_DIR)
```
I think at one point we changed from copy to symbolic link to avoid confusion when working with the files. Should this be similar?

adrimbarean commented on projects/iio_demo/src.mk:
```diff
@@ -86,3 +86,42 @@ INCS += $(PROJECT)/src/hal/stm32/main.h \
+ifeq (linux,$(PLATFORM))
+
+SRC_DIRS += $(NO-OS)/libraries/mbedtls/include/mbedtls
+SRCS += $(NO-OS)/network/linux_socket/linux_socket.c 
```
nitpick: maybe commit order should change here because linux_socket does not exist yet by this point.

adrimbarean commented on network/linux_socket/linux_socket.c:
```diff
@@ -0,0 +1,281 @@
+	int32_t ret;
+
+	ret = close(sock_id);
+
```
nitpick: extra line

adrimbarean commented on iio/iio_app/iio_app.c:
```diff
@@ -71,6 +77,9 @@ int32_t iio_app_run(struct iio_app_device *devices, int32_t len)
 	struct wifi_init_param		wifi_param;
 	struct wifi_desc		*wifi;
 	struct uart_desc		*uart_desc;
+#else
```
Probably should be `#elif defined (LINUX_PLATFORM)` because like this, if neither `LINUX_PLATFORM` or `USE_TCP_SOCKET` are defined, I think `struct tcp_socket_init_param` doesn't exist. CI signals this error.

mchindri commented on network/linux_socket/linux_socket.c:
```diff
@@ -0,0 +1,281 @@
+	socklen_t len = sizeof(saddr);
+
+	flags = fcntl(sock_id, F_GETFL);
+	fcntl(sock_id, F_SETFL, flags | O_NONBLOCK);
```
Is fcnlt needed here or you forgot to remove it?

ranechita commented:
v7

- fixed iio_app
- removed linux_desc and unused defines from linux_socket.h
- removed fcnlt from linux_socket_connect

mchindri reviewed (approved):
LGTM:
If @adrimbarean has no more comments and all checks pass I think it can be merged
